File: dev-client-api/algorithms/placement/shards.py
#!/usr/bin/env python3
"""
shards.py — Pure Python functions for shard packing.
"""

import math

def pack_rectangles(rectangles, gap):
    """
    Packs rectangles with a given spacing to fit in a roughly square bounding box (2D shelf packing).
    rectangles: list of dicts with 'id', 'w', 'd'
    gap: spacing between rectangles

    Returns: (width_used, depth_used, positions)
      positions: dict of id -> (u, v) representing bottom-left of the rectangle (including spacing)
    """
    if not rectangles:
        return 0, 0, {}

    # Sort by depth (d) descending to pack using a shelf algorithm
    sorted_rects = sorted(rectangles, key=lambda r: r['d'], reverse=True)

    # Calculate target width of the packed area:
    # We want it to be roughly square, so target_w is sqrt(sum_of_areas)
    total_area = sum((r['w'] + gap) * (r['d'] + gap) for r in sorted_rects)
    max_w = max(r['w'] + gap for r in sorted_rects)
    target_w = max(math.ceil(math.sqrt(total_area)), max_w)

    # Pack into shelves
    shelves = []  # each shelf: {"y_start": y, "height": h, "x_cursor": x}
    positions = {}

    for r in sorted_rects:
        rid = r['id']
        rw = r['w'] + gap
        rd = r['d'] + gap

        # Try to place in an existing shelf
        placed = False
        for shelf in shelves:
            if shelf["x_cursor"] + rw <= target_w:
                # Fits in this shelf
                positions[rid] = (shelf["x_cursor"], shelf["y_start"])
                shelf["x_cursor"] += rw
                shelf["height"] = max(shelf["height"], rd)
                placed = True
                break

        if not placed:
            # Create a new shelf
            y_start = 0
            if shelves:
                prev = shelves[-1]
                y_start = prev["y_start"] + prev["height"]

            new_shelf = {
                "y_start": y_start,
                "height": rd,
                "x_cursor": rw
            }
            positions[rid] = (0, y_start)
            shelves.append(new_shelf)

    # Calculate bounding box of packed area
    width_used = max(shelf["x_cursor"] for shelf in shelves) if shelves else 0
    depth_used = sum(shelf["height"] for shelf in shelves) if shelves else 0

    return width_used, depth_used, positions


def pack_shards_locally(dept_buckets, overrides_shards, gap_shards):
    """
    Packs shards locally within their respective departments.
    Returns: (dept_packings, dept_rects)
    """
    dept_packings = {}
    dept_rects = []

    for dept_name, dept_shards in dept_buckets.items():
        rects = []
        for s in dept_shards:
            s_key = f"{dept_name}.{s.name}" if hasattr(s, 'name') else s.get('key')
            # Extract names depending on whether we got objects or dicts
            s_name = s.name if hasattr(s, 'name') else s.get('shard')
            s_w = s.x if hasattr(s, 'x') else s.get('size', {}).get('w', 32)
            s_d = s.y if hasattr(s, 'y') else s.get('size', {}).get('d', 32)

            shard_override = overrides_shards.get(s_key, {})
            w = shard_override.get("size", {}).get("w", s_w)
            d = shard_override.get("size", {}).get("d", s_d)
            rects.append({"id": s_name, "w": w, "d": d})

        w_dept, d_dept, shard_positions = pack_rectangles(rects, gap_shards)
        dept_packings[dept_name] = {"w": w_dept, "d": d_dept, "positions": shard_positions}
        dept_rects.append({"id": dept_name, "w": w_dept, "d": d_dept})

    return dept_packings, dept_rects

File: dev-js-api/js/algorithms/placement/shards.js
/**
 * @fileoverview shards.js — Pure algorithms for packing shards inside departments.
 */

/**
 * 2D Shelf Packing Algorithm for rectangles.
 * Packs rectangles with a given spacing to fit in a roughly square bounding box.
 * 
 * @param {Array<{id: string, w: number, d: number}>} rectangles - Rectangles to pack
 * @param {number} gap - Spacing between rectangles
 * @returns {{widthUsed: number, depthUsed: number, positions: Object<string, {u: number, v: number}>}}
 */
export function packRectangles(rectangles, gap) {
  if (!rectangles || rectangles.length === 0) {
    return { widthUsed: 0, depthUsed: 0, positions: {} };
  }

  // Sort by depth descending (shelf algorithm heuristic)
  const sortedRects = [...rectangles].sort((a, b) => b.d - a.d);

  const totalArea = sortedRects.reduce((sum, r) => sum + (r.w + gap) * (r.d + gap), 0);
  const maxW = Math.max(...sortedRects.map(r => r.w + gap));
  const targetW = Math.max(Math.ceil(Math.sqrt(totalArea)), maxW);

  const shelves = [];
  const positions = {};

  for (const r of sortedRects) {
    const rid = r.id;
    const rw = r.w + gap;
    const rd = r.d + gap;

    let placed = false;
    for (const shelf of shelves) {
      if (shelf.xCursor + rw <= targetW) {
        positions[rid] = { u: shelf.xCursor, v: shelf.yStart };
        shelf.xCursor += rw;
        shelf.height = Math.max(shelf.height, rd);
        placed = true;
        break;
      }
    }

    if (!placed) {
      let yStart = 0;
      if (shelves.length > 0) {
        const prev = shelves[shelves.length - 1];
        yStart = prev.yStart + prev.height;
      }

      const newShelf = {
        yStart: yStart,
        height: rd,
        xCursor: rw
      };
      positions[rid] = { u: 0, v: yStart };
      shelves.push(newShelf);
    }
  }

  const widthUsed = shelves.length > 0 ? Math.max(...shelves.map(s => s.xCursor)) : 0;
  const depthUsed = shelves.reduce((sum, s) => sum + s.height, 0);

  return { widthUsed, depthUsed, positions };
}

/**
 * Packs shards locally within their respective departments.
 * 
 * @param {Object<string, Array>} deptBuckets - Shards grouped by department name
 * @param {Object} overridesShards - Overrides dictionary for shards
 * @param {number} gapShards - Gap spacing between shards
 * @returns {{deptPackings: Object, deptRects: Array}} Local packings and department bounds
 */
export function packShardsLocally(deptBuckets, overridesShards, gapShards) {
  const deptPackings = {};
  const deptRects = [];

  Object.entries(deptBuckets).forEach(([deptName, deptShards]) => {
    const rects = deptShards.map(s => {
      const shardOverride = overridesShards[s.key] || {};
      const w = shardOverride.size?.w !== undefined ? shardOverride.size.w : s.size.w;
      const d = shardOverride.size?.d !== undefined ? shardOverride.size.d : s.size.d;
      return { id: s.shard, w, d };
    });

    const { widthUsed, depthUsed, positions } = packRectangles(rects, gapShards);
    deptPackings[deptName] = { w: widthUsed, d: depthUsed, positions };
    deptRects.push({ id: deptName, w: widthUsed, d: depthUsed });
  });

  return { deptPackings, deptRects };
}
